### DISCUSSION TITLE
Order of `before` decorations in `TextEditor` placed at equal position

### DISCUSSION THREAD

**Discussion by vaclavHala** [microsoft/vscode-discussions#1025] [Extension Development QnA] [answered]
I'm trying to create a source code annotation which can have multiple "columns" of data for every source line. Result should look something like this
![before_decorations](https://github.com/microsoft/vscode-discussions/assets/12083193/ce7c4b86-2e7f-43de-8adc-478f653223c8)
Each column is composed of cells and each cell is a separate `before` part of `TextDecoration` placed on each line. The code looks something like this
```js
const decorPositions: vscode.DecorationOptions[] = []
// iterate over the code and for each line add two decorations like this
decorPositions.push({
  // so there will be two decorations in the array with this exact range
  range: line.range.with({end:line.range.start}), 
  renderOptions: { before: { 
    contentText: decorLabel(),
    backgroundColor: decorColor(),
    ...decorOpts 
  } }
})

ed.setDecorations(decorType, decorPositions)
```
Problem is vscode does not respect the order of decorations in the array but renders the `before` parts in arbitrary order. This is a problem as sometimes the cell containing data for column A ends up in position B and vice versa.

Only way I found to circumvent this problem is modifying the `range.start` of the second decoration on each line like this
```
start:line.range.start.with(line.range.start.with({character: 1}))
```
which however is very undesirable as it makes it possible for cursor to move "between" the two annotation columns.

Is there some way I can force vscode to render the before parts in the order I need?

**Comment by isidorn** [comment]
@hediet should be able to help here

**Comment by abrilurena** [comment]
Hi, @vaclavHala! I hope you're doing great. It's been a while, so I was wondering - do you still need help with this?

**Reply by vaclavHala** [reply]
Hello, by trial and error we discovered if we use different `TextEditorDecorationType` for each of the "columns" 
```js
const col1Type = vscode.window.createTextEditorDecorationType(...)
const col2Type = vscode.window.createTextEditorDecorationType(...)
...
const col1Positions: vscode.DecorationOptions[] = []
const col2Positions: vscode.DecorationOptions[] = []
// iterate over the code and for each line add one decoration per column into one of the `col*Positions` array

ed.setDecorations(col1Type, col1Positions)
ed.setDecorations(col2Type, col2Positions)
```
the order is always preserved. But this is not actually documented anywhere (that I could find).

If you could confirm this behavior is intentional and the API will behave this way also in the future that would be great (maybe mention it in the docs ;) ).
Otherwise if this behavior is undefined and the implementation just happens to behave this way at the moment, I'd like to hear that also. Of course if there is other way to achieve the same result that I missed that is the "correct" approach in this case please let me know.

**Reply by abrilurena** [reply]
@vaclavHala - Hi! I hope you're doing great. thank you for your patience ✨ I'm adding @isidorn as he might have more insights about it :D

**Reply by isidorn** [reply]
@hediet should be able to help confirm this API behaivour.

**Reply by burkeholland** [reply, marked as answer]
Per @hediet - these advanced decorations are not officially supported in the editor. So even if you're able to get this working, we can't guarantee that it won't break at some point in the future due to changes made in VS Code.

**Reply by alefragnani** [reply]
Even not being officially supported, I suppose this is the way to go. I had to apply the same technique on my Bookmarks extension (for different reasons) based on @alexdima suggestion in https://github.com/microsoft/vscode/issues/158837#issuecomment-1276757982. 

Also, it seems reasonable to think that the order extensions applies their decorators to a range should be respected, otherwise you could end up with different extensions fighting each other while applying similar decorators. Wouldn't it?

Maybe a solution could be to improve the Decorators API with something similar to the `priority` property available in the `createStatusBarItem` API. 

```typescript
         /**
         * Creates a status bar {@link StatusBarItem item}.
         *
         * @param id The unique identifier of the item.
         * @param alignment The alignment of the item.
         * @param priority The priority of the item. Higher values mean the item should be shown more to the left.
         * @return A new status bar item.
         */
        export function createStatusBarItem(id: string, alignment?: StatusBarAlignment, priority?: number): StatusBarItem;
```

Not sure this is doable, but if yes, a feature request could be filled in the VS Code repo https://github.com/microsoft/vscode/issues/new/choose
